core/node/da_dispatcher/src/da_dispatcher.rs

```diff
@@ -27,178 +36,296 @@ impl DataAvailabilityDispatcher {
         config: DADispatcherConfig,
         client: Box<dyn DataAvailabilityClient>,
     ) -> Self {
+        let request_semaphore = Arc::new(tokio::sync::Semaphore::new(100));
         Self {
             pool,
             config,
             client,
+            request_semaphore,
         }
     }
 
-    pub async fn run(self, mut stop_receiver: Receiver<bool>) -> anyhow::Result<()> {
-        loop {
-            if *stop_receiver.borrow() {
-                break;
-            }
+    pub async fn run(self, stop_receiver: Receiver<bool>) -> anyhow::Result<()> {
+        let subtasks = futures::future::join(
+            async {
+                if let Err(err) = self.dispatch_batches(stop_receiver.clone()).await {
+                    tracing::error!("dispatch error {err:?}");
+                }
+            },
+            async {
+                if let Err(err) = self.inclusion_poller(stop_receiver.clone()).await {
+                    tracing::error!("poll_for_inclusion error {err:?}");
+                }
+            },
+        );
+
+        tokio::select! {
+            _ = subtasks => {},
+        }
+        Ok(())
+    }
+
+    async fn dispatch_batches(&self, stop_receiver: Receiver<bool>) -> anyhow::Result<()> {
+        let (tx, mut rx) = mpsc::channel(100);
+
+        let next_expected_batch = Arc::new(AtomicI64::new(NO_FIRST_BATCH));
```

Comment:
can you add some docs here about why this atomic is needed?

Reply:
I've replaced the `AtomicI64` in favour of a `Mutex` that holds an `Option`, which makes the code clearer. I can add some docs if you still consider it necessary.